packages/kit/test/utils.js

```diff
@@ -143,11 +126,34 @@ export const test = base.extend({
 			});
 		}
 
-		use(async () => {
-			if (javaScriptEnabled) {
-				await page.waitForFunction(() => window.started);
-			}
-		});
+		const goto = page.goto;
+		page.goto =
+			/**
+			 * @param {string} url
+			 * @param {object}	opts
+			 */
+			async function (url, opts) {
+				const res = await goto.call(page, url, opts);
+				if (javaScriptEnabled) {
+					await page.waitForFunction(() => window.started);
+				}
+				return res;
+			};
+		await use(page);
```

Comment:
I'm not sure how safe this is to override `goto`, but I think this should be something like:

```js
const promises = []
if (javascriptEnabled) {
  promises.push(page.waitForFunction(() => window.started))
}
promises.push(goto.call(page, url, opts))
return Promise.all(promises)[promises.length - 1]
```

It could be refactored to look nicer, but in my experience, the `waitFor` must always be called before the goto. If we apply this pattern for the `started()` code we use to have, it should fix the flakes too.

This `Promise.all` pattern is describe [here](https://playwright.dev/docs/navigations#asynchronous-navigation).

Reply:
Advisability of monkey-patching notwithstanding, I think it's fine? it shouldn't be racy (it's not like `waitForNavigation` where you need to start waiting before triggering the navigation), and you need to go to the page before checking for `window.started`, otherwise in a situation like this...

```js
await page.goto('/foo');
expect(x).toBe(y);

await page.goto('/bar'); // `window.started` is still true from before
expect(x).toBe(y);
```

...the promise could resolve prematurely, I think?

Reply:
Yeah, now that I think of it, since `started` is a promise we could wait on that later as well, so the race condition I'm imagining wouldn't be an issue. And the example is true too. LGTM then 👍